src/Sarif/Query/Evaluators/PropertyBagPropertyEvaluator.cs

```diff
@@ -0,0 +1,108 @@
+
+        public PropertyBagPropertyEvaluator(TermExpression term)
+        {
+            Match match = s_propertyNameRegex.Match(term.PropertyName);
+            _propertyName = match.Groups["name"].Value;
+
+            string prefix = match.Groups["prefix"].Value;
+            if (prefix.Equals(RulePropertyPrefix))
+            {
+                _propertyBelongsToRule = true;
+            }
+            else if (!prefix.Equals(ResultPropertyPrefix))
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        SdkResources.ErrorInvalidQueryPropertyPrefix,
+                        "'" + string.Join("', '", ResultPropertyPrefix, RulePropertyPrefix)) + "'",
+                    nameof(term));
+            }
+
+            _evaluator = CreateEvaluator(term);
+        }
+
+        // Create an appropriate evaluator object for the specified term. For operators that apply
+        // only to strings (":" (contains), "|>" (startswith), or ">|" (endswith), always create
+        // a string evaluator.  All other operators (==, !=, >, <, etc.) can apply to both strings
+        // and numbers. If the value being compared parses as a number, assume that a numeric
+        // comparison was intended, and create a numeric evaluator. Otherwise, create a string evaluator.
+        // This could cause problems if the comparand is string that happens to look like a number.
+        private IExpressionEvaluator<Result> CreateEvaluator(TermExpression term) =>
+           IsStringComparison(term)
+            ? new StringEvaluator<Result>(GetProperty<string>, term, StringComparison.OrdinalIgnoreCase) as IExpressionEvaluator<Result>
+            : new DoubleEvaluator<Result>(GetProperty<double>, term);
+
+        private static readonly ReadOnlyCollection<CompareOperator> s_stringSpecificOperators =
+            new ReadOnlyCollection<CompareOperator>(
+                new CompareOperator[]
+                {
+                    CompareOperator.Contains,
+                    CompareOperator.EndsWith,
+                    CompareOperator.StartsWith
+                });
+
+        private bool IsStringComparison(TermExpression term)
+            => s_stringSpecificOperators.Contains(term.Operator) || !double.TryParse(term.Value, out _);
+
+        private T GetProperty<T>(Result result)
+        {
+            PropertyBagHolder holder = GetPropertyBagHolder(result);
+            return GetPropertyFromHolder<T>(holder);
+        }
+
+        private PropertyBagHolder GetPropertyBagHolder(Result result) =>
+            _propertyBelongsToRule
+            ? result.GetRule() as PropertyBagHolder
+            : result;
+
+        private T GetPropertyFromHolder<T>(PropertyBagHolder holder)
+        {
+            try
+            {
+                return holder.TryGetProperty(_propertyName, out T value) ? value : default;
```

Comment:
>TryGetProperty [](start = 30, length = 14)

this looks weird. Why is this named TryGetProperty if it can raise an exception? seems wrong to me. instead, TryProperty should handle the exception and return 'false'.

i won't block on this but i don't this it's right.